### PR TITLE
Add RGB color system

### PR DESCRIPTION
**Pull request by alppp**
adds full RGB color support using the §x§R§R§G§G§B§B wire format (1.16+ spigot standard).
degrades gracefully on unmodded clients.

features:
- 16.7M RGB colors via §x or &#RRGGBB shorthand
- gradients (§g/&g) with two color endpoints
- rainbow (§y/&y) - static per-character hue cycle
- wave (§w/&w) - animated sine offset, freezes in HUD cache
- dinnerbone (§j/&j) - flipped characters
- all legacy §0-f and §k-o still work
- line wrapping preserves colors and effects across breaks
- config toggles: enableRGBColors, enableTextEffects, enableAmpersandConversion
- sign rendering support

Examples :
```
--- Basic RGB (&#RRGGBB syntax) ---
&#FF0000Red &#00FF00Green &#0000FFBlue

--- Legacy colors ---
&cRed &aGreen &bBlue &eYellow

--- Rainbow ---
&qRainbow text here&r normal

--- Wave ---
&zWavy text here&r normal

--- Dinnerbone ---
&vFlipped text&r normal

--- Gradient ---
&g&#FF0000&#0000FFSmooth red to blue gradient&r done

--- Wave survives color change ---
&z&#FF0000Red wave &aGreen wave still going&r stopped

--- Dinnerbone survives color change ---
&v&cFlipped red &aFlipped green &bFlipped blue&r normal

--- Reset clears everything ---
&z&q&vAll effects on&r All effects off

--- Gradient + bold ---
&g&#FF0000&#FFFF00&lBold gradient text&r done

--- Invalid edge cases (should render literally) ---
&#GGGGGGinvalid hex &#FFnot enough digits &# nothing

--- Normal formatting ---
Just a normal &c with &l bold &r reset

--- Long gradient wrap test ---
&g&#FF0000&#00FF00This is a really long gradient message that should wrap across multiple chat lines to verify the color interpolation continues smoothly after the line break point

```
image

related PRs (all required):
- https://github.com/GTNewHorizons/GTNHLib/pull/328 - & code awareness in width/size/trim methods
- https://github.com/GTNewHorizons/Hodgepodge/pull/829 - chat wrap, text field scroll, anvil limit, sign input
- https://github.com/GTNewHorizons/BetterQuesting/pull/199 - editor buttons, color picker, gradient wrap, monochrome fix

### CODE REVIEW

File: src/mixin/java/com/gtnewhorizons/angelica/mixins/early/angelica/fontrenderer/MixinFontRenderer.java [hunk 1]
```diff
@@ -120,6 +124,15 @@ public abstract class MixinFontRenderer implements FontRendererAccessor, IFontPa
...
         boolean unicodeMode, CallbackInfo ci) {
         angelica$batcher = new BatchingFontRenderer((FontRenderer) (Object) this, this.charWidth, this.colorCode, this.locationFontTexture);
+        try {
+            FontRendering.setTextPreprocessor(text ->
+                AngelicaConfig.enableAmpersandConversion
+                    ? ColorCodeUtils.convertAmpersandToSectionX(text)
+                    : text
+            );
+        } catch (NoSuchMethodError ignored) {
+            // Old GTNHLib without text preprocessor — & conversion only at render time
```
Comment (mitchej123) [inline src/mixin/java/com/gtnewhorizons/angelica/mixins/early/angelica/fontrenderer/MixinFontRenderer.java]: Catch is slow, especially in a hot path.  I'd suggest doing some version/class checks early on in startup and set whether it's available or not.  Then using a compat class, only call if that boolean is true.

File: src/main/java/com/gtnewhorizons/angelica/client/font/BatchingFontRenderer.java [hunk 2]
```diff
@@ -616,23 +762,46 @@ public float drawString(final float anchorX, final float anchorY, final int colo
...
 
                 if (curBold) {
                     for (int n = 1; n <= boldCopies; n++) {
                         final float shadowOffsetPart = shadowOffset * ((float) n / boldCopies);
-                        pushTexRect(curX + shadowOffsetPart, heightNorth, glyphW - 1.0f, heightSouth, itOff, curColor, uStart, vStart, uSz, vSz);
+                        if (curDinnerbone) {
+                            pushTexRectVFlipped(curX + shadowOffsetPart, renderY, glyphW - 1.0f, heightSouth, itOff, curColor, uStart, vStart, uSz, vSz);
+                        } else {
+                            pushTexRect(curX + shadowOffsetPart, renderY, glyphW - 1.0f, heightSouth, itOff, curColor, uStart, vStart, uSz, vSz);
+                        }
```
Comment (DeathFuel) [inline src/main/java/com/gtnewhorizons/angelica/client/font/BatchingFontRenderer.java]: Is there any nicer way of doing this, without duplicating every call? Perhaps passing `curDinnerbone` as a parameter that flips V inside of `pushTexRect` would work (if there's no better way, like a mathematical transformation along with some way to avoid tex bounds issues)?

File: src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java [hunk 3]
```diff
@@ -107,6 +107,18 @@ public class AngelicaConfig {
...
+    @Config.Comment("Enable RGB color support via §x§R§R§G§G§B§B format codes")
+    @Config.DefaultBoolean(true)
+    public static boolean enableRGBColors;
+
+    @Config.Comment("Enable text effects: rainbow (§y), wave (§w), gradient (§g), dinnerbone (§j)")
+    @Config.DefaultBoolean(true)
+    public static boolean enableTextEffects;
+
+    @Config.Comment("Convert &RRGGBB to §x color codes at render time")
+    @Config.DefaultBoolean(true)
```
Comment (DeathFuel) [inline src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java:119]: Does this setting cause a problem where stuff like `PB&J sandwich` is affected by text formatting? If so, perhaps it shouldn't be true by default (though the bug reports would be funny for sure).

File: src/mixin/java/com/gtnewhorizons/angelica/mixins/early/angelica/fontrenderer/MixinFontRenderer.java (hunk 1 above)
Comment (alppp) [reply src/mixin/java/com/gtnewhorizons/angelica/mixins/early/angelica/fontrenderer/MixinFontRenderer.java]: Replaced with a compat class (`GTNHLibCompat`) that does a one-time reflection check in a static initializer. The mixin init now just checks a boolean.

File: src/main/java/com/gtnewhorizons/angelica/client/font/BatchingFontRenderer.java (hunk 2 above)
Comment (alppp) [reply src/main/java/com/gtnewhorizons/angelica/client/font/BatchingFontRenderer.java]: Added a `boolean flipV` parameter to `pushTexRect` computes vTop/vBot from the flag and swaps V coordinates when true.

File: src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java (hunk 3 above)
Comment (alppp) [reply src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java:119]: The `&#RRGGBB` syntax (with hash prefix) is specifically designed to avoid this `PB&J` does NOT match because there's no `#` after `&`. (technically it did match with dinnerbone effect so below : )
However, the single-letter effect codes (`&y` rainbow, `&w` wave, `&j` dinnerbone, `&g` gradient) were collision-prone with natural text (`&yes`, `&with`, `PB&J`, `&good`). Remapped them to low-collision letters : &q, &v, &z.

**Comment by Dream-Master** [comment]
@alppp a dependent pr is already merged will this cause problem if Angelica pr is not merged? 
@mitchej123 maybe you can have a look whenever you found time for it ?

**Comment by alppp** [comment]
- split enableTextEffects into individual per-effect toggles (enableGradients, enableRainbow, enableWaveText, enableDinnerboneText)  all default on
- fix wave text frequency being higher on formatted strings (was using raw charIdx including format codes, now uses visible char count)
- add "Text Effects" options page in video settings with toggles for each effect + ampersand conversion
- added amplitude slider to wavy
image

https://github.com/user-attachments/assets/f4fd0316-19e6-4cd5-b5ea-f17766745f15

File: src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java [hunk 4]
```diff
@@ -111,9 +111,26 @@ public class AngelicaConfig {
...
+    public static boolean enableWaveText;
+
+    @Config.Comment("Enable upside-down text (&v)")
+    @Config.DefaultBoolean(true)
+    public static boolean enableDinnerboneText;
+
+    @Config.Comment("Wave text amplitude (how far characters bounce)")
+    @Config.DefaultFloat(2.0f)
+    @Config.RangeFloat(min = 0.5f, max = 8.0f)
```
Comment (DeathFuel) [inline src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java]: Minimum value of 0.5

File: src/main/java/me/jellysquid/mods/sodium/client/gui/SodiumGameOptionPages.java [hunk 5]
```diff
@@ -534,4 +534,62 @@ public static OptionPage appearance() {
...
+                        .setName("Wave Text")
+                        .setTooltip("Animated bouncing text via &z")
+                        .setControl(TickBoxControl::new)
+                        .setBinding((opts, value) -> AngelicaConfig.enableWaveText = value,
+                                    opts -> AngelicaConfig.enableWaveText)
+                        .build())
+                .add(OptionImpl.createBuilder(int.class, angelicaOpts)
+                        .setName("Wave Amplitude")
+                        .setTooltip("How far wave text bounces (default 2)")
+                        .setControl(option -> new SliderControl(option, 1, 8, 1, ControlValueFormatter.number()))
```
Comment (DeathFuel) [inline src/main/java/me/jellysquid/mods/sodium/client/gui/SodiumGameOptionPages.java]: Minimum value of 1

**Review by DeathFuel — COMMENTED** [review]
The min values for the wave amplitude setting are inconsistent.

**Review by DeathFuel — APPROVED** [review]
Hand it off to the beta testers...

**Review by mitchej123 — CHANGES_REQUESTED** [review]
There's a few allocation/efficiency concerns here.  I'll get to them, just marking this as requested changes so someone else doesn't merge it.

**Comment by mitchej123** [comment]
@alppp Try with my push.  If you're happy give a :+1: and we'll merge it.

**Comment by alppp** [comment]
I tested, looks perfect. Is good to go!

**Review by mitchej123 — APPROVED** [review]
